USACO Silver 2022 Range Reconstruction (#6248)
cpinitiative/usaco-guide: solutions/silver/usaco-1256.mdx

---
id: usaco-1256
source: USACO Silver 2022 December
title: Range Reconstruction
author: Aarav Sethi
---

> 

Reviewer: official editorial?

## Explanation

> ### Subtask 1: $\mathcal{O}(N^2 \cdot 2^N)$

Reviewer: ```suggestion
The first observation is that we can query two adjacent indices $i$ and $j=i+1$ and be given the absolute difference $|a_i-a_j|$. If we set $a_0$ to an arbitrary value, we have $a_1 = a_0 \pm r_{0,1}, a_2 = a_1 \pm r_{1, 2}, \dots, a_n = a_{n-1} \pm r_{n-1,n}$. 

For this subtask it suffices to simply brute force the signs of each $r_{i,j}$ using bitmasking or recursion, and checking each possible candidate.
```

The first observation is that you may query two adjacent indices $i$ and $j=i+1$ and be given the absolute difference $|a_i-a_j|$.

If you pick an arbitrary $a_0$ and use the above query format for every $0\leq i\leq n-2$ in order, you would have two candidates ($x-a_i$ and $a_i-x$) each time you query.

For this subtask it suffices to simply brute force all combinations of candidates at each step..

> #### Example snippet:

Reviewer: i think this is better in a spoiler

---
id: usaco-1256
source: USACO Silver 2022 December
title: Range Reconstruction
author: Aarav Sethi
---

## Explanation
### Subtask 1: $\mathcal{O}(N^2 \cdot 2^N)$

> The first observation you can make is that because you can query for two adjacent indices $i$ and $j$ you will be given the absolute difference $|a_i-a_j|$.

Reviewer 2: Cut the jargon/reptition; also(n) show $j=i+1$ explicitly (or just get rid of $j$ and state $i+1$).  Example:
```suggestion
The first observation is that you may query two adjacent indices $i$ and $j=i+1$ and be given the absolute difference $|a_i-a_j|$.
```
I suggest rewriting this on your own, as my writing skills are kinda homeless.

> If you perform a left to right sweep at every index you would have two candidates ($x-a_i$ and $a_i-x$).

Reviewer 2: Make this explicit and cut the paragraph break.  Again I suggest rewriting this on your own instead of using my writing.
```suggestion

If you pick an arbitrary $a_0$ and use the above query format for every $0\leq i\leq n-2$ in order, you would have two candidates ($x-a_i$ and $a_i-x$) each time you query.
```

> For this subtask it would be satisfactory to simply brute force all combinations of candidates to use.

Reviewer 2: ```suggestion
For this subtask it suffices to simply brute force all combinations of candidates at each step..
```
See above disclaimer

> > **Note:** $a_0$ can become any arbitrary integer, but it is important to make sure that none of the values of the array go negative. This can be ensured while printing by adding $min(0, -minVal)$ to every element.

Reviewer 2: Is this even required by the problem?  From problem:

> The values in your array should be in the range $[-10^9, 10^9]$.

So I don't see how you have to be positive.

If it is required:
- Don't mention $a_0$ being arbitrary here, mention it earlier
- Define `minVal`

Author: Good catch thank you! I have no idea why I thought numbers were restricted to be $> 0$.

Reviewer 2: Welcome!  Stil you admit orz

#### Example snippet:
<LanguageSection>

<CPPSection>

```cpp
for (ll i = 0; i < (1 << (n - 1)); i++) {
	// Generate solution based on bitmask on whether to add or subtract at each pos
	solution[0] = 1;
	for (int j = 1; j < n; j++) {
		if (i & (1 << (j - 1))) solution[j] = solution[j - 1] + r[j - 1][j];
		else solution[j] = solution[j - 1] - r[j - 1][j];
	}

	// Test all rules
	bool valid = true;
	for (int x = 0; x < n; x++) {
		int maxVal = solution[x], minVal = solution[x];
		for (int y = x + 1; y < n; y++) {
			maxVal = max(maxVal, solution[y]);
			minVal = min(minVal, solution[y]);

			if (r[x][y] != maxVal - minVal) {
				valid = false;
				break;
			}
		}

		if (!valid) break;
	}
}
```

### Subtask 2: $\mathcal{O}(N^2)$

> One more observation that can be made, is that a greedy algorithm is guarenteed to be correct. This is because backtracking would never need to be used since at each point you could check each of the candidates and if they would pass all previous checks.

Reviewer 2: Why?  Give either some intuition or proof on why we can just sweep once, perhaps illustrate with an example.

Reviewer: yeah agreed, i think editorial says proof by induction so you could prob explain some of that

(i'm too lazy to figure out the proof rn(

> The optimized algorithm would involve again going left to right and generating two candidates, but instead of basing which candidates to use on bruteforce it would pick whichever is currently valid.

Reviewer 2: Redundant / probably can merge into previous paragraph

</CPPSection>

> </LanguageSection>

Reviewer 2: The two closing tags should go before Subtask 2.

Reviewer 2: Discriminating Python and Java...

Reviewer 2: FWIW not saying you intetntionally didn't address them but these comments are unaddressed despite being marked as outdated as some of the context change.d

Reviewer 2: @AsynchronousAI Nevermind here... was looking at wrong revision.  This is resolved.

## Implementation

**Time Complexity:** $\mathcal{O}(N^2)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>

using ll = long long;
using namespace std;

int main() {
	// Read input
	int n;
	cin >> n;
	vector<vector<int>> r(n, vector<int>(n));
	for (int i = 0; i < n; i++)

> 		for (int j = i; j < n; j++) cin >> r[i][j];

Reviewer 2: Braces(on) :)

Reviewer 2: I just added the phrase "son" to "braces".

Use braces.
